packages/http/src/merge-patch.ts

```diff
@@ -0,0 +1,465 @@
+                        ? "@statusCode"
+                        : undefined;
+              if (decorator) {
+                reportDiagnostic(program, {
+                  code: "merge-patch-contains-metadata",
+                  target: prop,
+                  format: { metadataType: decorator, propertyName: prop.name },
+                });
+              }
+            }
+          }
+
+          clone.decorators = clone.decorators.filter((d) => d.decorator !== $applyMergePatch);
+          ctx.program.stateMap(HttpStateKeys.mergePatchModel).set(clone, model);
+          // ctx.call($friendlyName, clone, nameTemplate, clone);
+          rename(ctx.program, clone, nameTemplate);
+        },
+      },
+      ModelProperty: {
+        filter: () => MutatorFlow.DoNotRecur,
+        mutate: (prop, clone, program) => {
+          if (isMergePatchSubject(prop.type)) {
+            clone.type = cachedMutateSubgraph(
+              program,
+              prop.optional ? (_optionalInteriorMutator ?? self) : self,
+              prop.type,
+            ).type;
+          }
+          ctx.program.stateMap(HttpStateKeys.mergePatchProperty).set(clone, prop);
+        },
+      },
+      UnionVariant: {
+        filter: () => MutatorFlow.DoNotRecur,
+        mutate: (variant, clone, program) => {
+          if (isMergePatchSubject(variant.type)) {
+            const mutated = cachedMutateSubgraph(
+              program,
+              _optionalInteriorMutator || self,
+              variant.type,
+            );
+            clone.type = mutated.type;
+          }
+        },
+      },
+      Tuple: {
+        filter: () => MutatorFlow.DoNotRecur,
+        mutate: (tuple, clone, program) => {
+          for (const [index, element] of tuple.values.entries()) {
+            if (isMergePatchSubject(element)) {
+              clone.values[index] = cachedMutateSubgraph(
+                program,
+                _replaceInteriorMutator ?? self,
+                element,
+              ).type;
+            }
+          }
+        },
+      },
+    };
+
+    return self;
+  }
+
+  function nullable(realm: Realm, t: Type): Union {
+    return $(realm).union.create({
+      variants: [
+        $(realm).unionVariant.create({
+          type: t,
+        }),
+        $(realm).unionVariant.create({
+          type: $(realm).intrinsic.null,
+        }),
+      ],
+    });
+  }
+}
+
+type MergePatchSubject = Model | Union | ModelProperty | UnionVariant | Tuple;
+
+function isMergePatchSubject(type: Type): type is MergePatchSubject {
+  return (
+    type.kind === "Model" ||
+    type.kind === "Union" ||
+    type.kind === "ModelProperty" ||
+    type.kind === "UnionVariant" ||
+    type.kind === "Tuple"
+  );
+}
+
+function rename(
+  program: Program,
+  type: Extract<MergePatchSubject, { name?: string }>,
+  nameTemplate: string,
+) {
+  // workaround for openapi3/asset-emitter issue that does not recognize renamed array literals
```

Comment:
whats the issue here? can we file a bug?

Reply:
Yes, will file a bug.  Since the array literal type always has name `Array`, it doesn't make much sense to rename it anyway.

Reply:
https://github.com/microsoft/typespec/issues/7226